src/lib.rs

```diff
@@ -51,137 +51,94 @@
-        (self.0.get() as usize) - 1
-    }
-}
+mod symbol;
 
-impl Symbol for usize {
-    fn from_usize(val: usize) -> Self {
-        val
-    }
+pub use crate::symbol::{
+    DefaultSymbol,
+    Symbol,
+};
+use cfg_if::cfg_if;
+use core::{
+    hash::{
+        BuildHasher,
+        Hash,
+        Hasher,
+    },
+    iter,
+    iter::FromIterator,
+    marker,
+    pin::Pin,
+    ptr::NonNull,
+    slice,
+};
 
-    fn to_usize(self) -> usize {
-        self
+cfg_if! {
+    if #[cfg(feature = "std")] {
+        use std::{
+            collections::{
+                hash_map::RandomState,
+                HashMap,
+            },
+            vec,
+        };
+    } else {
+        extern crate alloc;
+        use alloc::{
+            collections::{
+                btree_map::BTreeMap,
+            },
+            vec,
+        };
     }
 }
 
-/// Internal reference to `str` used only within the `StringInterner` itself
-/// to encapsulate the unsafe behaviour of interior references.
+/// Internal reference to an interned `str`.
+///
+/// This is a self-referential from the interners string map
+/// into the interner's actual vector of strings.
 #[derive(Debug, Copy, Clone, Eq)]
-struct InternalStrRef(*const str);
+struct PinnedStr(NonNull<str>);
 
-impl InternalStrRef {
-    /// Creates an InternalStrRef from a str.
-    ///
-    /// This just wraps the str internally.
+impl PinnedStr {
+    /// Creates a new `PinnedStr` from the given `str`.
     fn from_str(val: &str) -> Self {
-        InternalStrRef(val as *const str)
+        PinnedStr(NonNull::from(val))
     }
 
-    /// Reinterprets this InternalStrRef as a str.
-    ///
-    /// This is "safe" as long as this InternalStrRef only
-    /// refers to strs that outlive this instance or
-    /// the instance that owns this InternalStrRef.
-    /// This should hold true for `StringInterner`.
-    ///
-    /// Does not allocate memory!
-    fn as_str(&self) -> &str {
-        unsafe { &*self.0 }
+    /// Creates a new `PinnedStr` from the given pinned `str`.
+    fn from_pin(pinned: Pin<&str>) -> Self {
+        PinnedStr(NonNull::from(&*pinned))
     }
-}
 
-impl<T> From<T> for InternalStrRef
-where
-    T: AsRef<str>,
-{
-    fn from(val: T) -> Self {
-        InternalStrRef::from_str(val.as_ref())
+    /// Returns a shared reference to the underlying `str`.
+    fn as_str(&self) -> &str {
+        // SAFETY: This is safe since we only ever operate on interned `str`
+        //         that are never moved around in memory to avoid danling
+        //         references.
+        unsafe { self.0.as_ref() }
     }
 }
 
-impl Hash for InternalStrRef {
+impl Hash for PinnedStr {
     fn hash<H: Hasher>(&self, state: &mut H) {
         self.as_str().hash(state)
     }
 }
 
-impl PartialEq for InternalStrRef {
-    fn eq(&self, other: &InternalStrRef) -> bool {
+impl PartialEq for PinnedStr {
+    fn eq(&self, other: &Self) -> bool {
         self.as_str() == other.as_str()
```

Comment:
Why do you need to dereference the pointer here? I thought `string-interner` guarentees that two strings will always return the same Symbol, so shouldn't they have the same pointer as well?

Reply:
Good question! You can generally treat same symbols as being the same string. However, for the string interners internal hash map you need to compare incoming strings using their contents since they will probably not point to strings stored in the string interner and thus will have different pointers to potentially equal string contents.

Reply:
Ok, so a `PinnedStr` may not yet have been deduplicated? That makes sense.